File: tessera/Tessera/Lyrics.swift
// The words, fetched and lettered, phone-side.
//
// Same source and same behaviour as brain/art/lyrics.py: LRCLIB by track,
// the current line only, words arriving one at a time across the front of
// the line's window, each with a short brightness ramp. The pixel font is
// ASCII, and half the world's music is not: characters the font cannot
// letter step aside, so a mixed-script line keeps its English on time and a
// line with nothing letterable shows the sleeve alone for its bar.

import Foundation
import MediaPlayer
import QuartzCore

/// The song's position as a clock worth trusting.
///
/// MPMusicPlayerController.currentPlaybackTime is the truth, delivered
/// badly: it crosses an XPC bridge, staircases, and can repeat itself for
/// most of a second before jumping. Rendering lyrics straight off it is why
/// nothing ever felt locked. This keeps its own monotonic clock anchored to
/// the player's samples: between samples time flows smoothly at 1x; when a
/// fresh sample disagrees a little, the anchor is SLEWED gently toward it
/// (a fixed fraction per reading, never a visible jump); when it disagrees
/// a lot, that is a seek, and the clock jumps with it.
@MainActor
final class SongClock {
    private var anchorTime: Double = 0
    private var anchorHost: Double = 0
    private var lastRaw: Double = -1
    private var stateAt: Double = 0
    private var statePlaying = true
    private var music: MPMusicPlayerController { .systemMusicPlayer }

    func hardReset() { lastRaw = -1; stateAt = 0 }

    func now() -> Double {
        let raw = music.currentPlaybackTime
        let host = CACurrentMediaTime()
        // playbackState is a second XPC hop; it changes on the scale of
        // taps, not frames, so half a second of trust is plenty
        if host - stateAt > 0.5 {
            statePlaying = music.playbackState == .playing
            stateAt = host
        }
        guard statePlaying else {
            anchorTime = raw
            anchorHost = host
            lastRaw = raw
            return raw
        }
        if lastRaw < 0 {
            anchorTime = raw
            anchorHost = host
            lastRaw = raw
            return raw
        }
        let predicted = anchorTime + (host - anchorHost)
        if raw != lastRaw {
            lastRaw = raw
            let diff = raw - predicted
            if abs(diff) > 0.9 {
                // a seek, or the player was wrenched: follow it at once
                anchorTime = raw
                anchorHost = host
            } else {
                // a small disagreement: lean toward it, never lurch
                anchorTime += diff * 0.12
            }
        }
        return anchorTime + (host - anchorHost)
    }
}

/// One synced line: its start, its text, and, when the sheet is Enhanced
/// LRC, the actual moment each word is sung.
struct LyricLine {
    let t: Double
    let text: String
    let words: [(Double, String)]?
}

@MainActor
final class LyricsBook {
    private(set) var track = ""
    private(set) var sheet: [LyricLine]? = nil
    /// The second voice: ad-libs and echoes, each with the window it owns.
    private(set) var adlibs: [(start: Double, end: Double, text: String)] = []
    /// Supersession by generation, not a lock: the old design refused new
    /// asks while one was in flight, and a fetch that lost its race left the
    /// flag stuck forever — every later track then showed no words at all.
    private var generation = 0

    func ask(track id: String, artist: String, title: String, album: String,
             duration: Double? = nil) {
        guard id != track else { return }
        track = id
        sheet = nil
        generation += 1
        let gen = generation
        Task { [weak self] in
            let found = await Self.fetch(artist: artist, title: title,
                                         album: album, duration: duration)
            await MainActor.run {
                guard let self, self.generation == gen else { return }
                if let found {
                    let split = Self.splitVoices(found)
                    self.sheet = split.mains
                    self.adlibs = split.adlibs
                } else {
                    self.sheet = nil
                    self.adlibs = []
                }
            }
        }
    }

    /// Two voices out of one sheet. A line living entirely in parentheses
    /// is the second singer; a trailing parenthetical on a main line is an
    /// echo that belongs to that line's moment. Each ad-lib owns a window
    /// from its stamp to the next event, so it plays under whatever the
    /// first voice is doing.
    static func splitVoices(_ lines: [LyricLine])
        -> (mains: [LyricLine], adlibs: [(Double, Double, String)]) {
        var mains: [LyricLine] = []
        var raw: [(Double, String)] = []      // adlib starts, windows later
        let paren = /\([^()]*\)/
        for line in lines {
            let trimmed = line.text.trimmingCharacters(in: .whitespaces)
            let groups = trimmed.matches(of: paren).map { String($0.output) }
            for g in groups where g.count > 2 {
                raw.append((line.t, g))
            }
            let main = trimmed.replacing(paren, with: " ")
                .split(separator: " ").joined(separator: " ")
            if !main.isEmpty {
                // word stamps survive only if the line kept all its words
                mains.append(LyricLine(t: line.t, text: main,
                                       words: main == trimmed ? line.words : nil))
            }
        }
        let everyStart = lines.map { $0.t }.sorted()
        let adlibs = raw.map { (t, text) -> (Double, Double, String) in
            let next = everyStart.first { $0 > t + 0.05 } ?? (t + 4)
            return (t, min(next, t + 6), text)
        }
        return (mains, adlibs)
    }

    // MARK: fetch

    static func fetch(artist: String, title: String, album: String,
                      duration: Double? = nil) async -> [LyricLine]? {
        // nil = transport failed (worth retrying); [] = answered, no rows
        func request(_ path: String, _ items: [URLQueryItem]) async -> [[String: Any]]? {
            var comps = URLComponents(string: "https://lrclib.net/api/\(path)")!
            comps.queryItems = items
            guard let url = comps.url else { return [] }
            var req = URLRequest(url: url)
            req.timeoutInterval = 10
            guard let (data, resp) = try? await URLSession.shared.data(for: req) else {
                return nil
            }
            guard (resp as? HTTPURLResponse)?.statusCode == 200 else { return [] }
            if let one = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                return [one]
            }
            if let many = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] {
                return many
            }
            return []
        }

        // A network hiccup must not be remembered forever as "this song has
        // no words": transport errors retry with a breath between attempts.
        for attempt in 0..<3 {
            if attempt > 0 { try? await Task.sleep(for: .seconds(2 * attempt)) }

            var items: [URLQueryItem] = [
                .init(name: "artist_name", value: artist),
                .init(name: "track_name", value: title),
                .init(name: "album_name", value: album),
            ]
            if let duration { items.append(.init(name: "duration", value: String(Int(duration)))) }
            guard let direct = await request("get", items) else { continue }

            var candidates = direct.filter { $0["syncedLyrics"] is String }
            // The direct answer can still be the WRONG EDIT: a sheet synced
            // to the radio cut drifts a little more every chorus. When its
            // length disagrees with the song by more than a few seconds,
            // distrust it and search for a better one.
            if let d = duration, let hit = candidates.first,
               let hd = hit["duration"] as? Double, abs(hd - d) > 4 {
                candidates = []
            }
            if candidates.isEmpty {
                guard let found = await request("search", [
                    .init(name: "artist_name", value: artist),
                    .init(name: "track_name", value: title),
                ]) else { continue }
                candidates = found.filter { $0["syncedLyrics"] is String }
            }
            let best: [String: Any]? = duration.flatMap { d in
                candidates.min { a, b in
                    let da = abs((a["duration"] as? Double ?? 1e9) - d)
                    let db = abs((b["duration"] as? Double ?? 1e9) - d)
                    return da < db
                }
            } ?? candidates.first
            guard let raw = best?["syncedLyrics"] as? String else { return nil }
            let lines = parse(lrc: raw)
            return lines.isEmpty ? nil : lines
        }
        return nil
    }

    static func parse(lrc: String) -> [LyricLine] {
        // The whole spec, because the spec is where the sync lives:
        // [offset:±ms] is a shift baked in by whoever synced the file
        // (positive = the lyrics belong EARLIER), and <mm:ss.xx> stamps are
        // Enhanced LRC's real word times.
        var off = 0.0
        if let m = lrc.firstMatch(of: /\[offset:\s*([+-]?\d+)\s*\]/.ignoresCase()) {
            off = (Double(m.output.1) ?? 0) / 1000.0
        }
        var out: [LyricLine] = []
        let stamp = /\[(\d+):(\d+(?:\.\d+)?)\]/
        let wordStamp = /<(\d+):(\d+(?:\.\d+)?)>/
        for raw in lrc.split(separator: "\n", omittingEmptySubsequences: false) {
            let line = String(raw)
            let stamps = line.matches(of: stamp)
            guard !stamps.isEmpty else { continue }
            var text = line.replacing(stamp, with: "").trimmingCharacters(in: .whitespaces)
            var words: [(Double, String)]? = nil
            let wstamps = text.matches(of: wordStamp)
            if !wstamps.isEmpty {
                var built: [(Double, String)] = []
                var cursor = text.startIndex
                var pendingT: Double? = nil
                for m in wstamps {
                    let chunk = String(text[cursor..<m.range.lowerBound])
                        .trimmingCharacters(in: .whitespaces)
                    if let pt = pendingT, !chunk.isEmpty {
                        built.append((pt, chunk))
                    }
                    pendingT = (Double(m.output.1) ?? 0) * 60
                        + (Double(m.output.2) ?? 0) - off
                    cursor = m.range.upperBound
                }
                let tail = String(text[cursor...]).trimmingCharacters(in: .whitespaces)
                if let pt = pendingT, !tail.isEmpty { built.append((pt, tail)) }
                if !built.isEmpty {
                    words = built
                    text = built.map { $0.1 }.joined(separator: " ")
                }
            }
            for m in stamps {
                let t = (Double(m.output.1) ?? 0) * 60 + (Double(m.output.2) ?? 0) - off
                out.append(LyricLine(t: t, text: text, words: words))
            }
        }
        out.sort { $0.t < $1.t }
        var merged: [LyricLine] = []
        for line in out {
            if let last = merged.last, line.t - last.t < 0.05 {
                let joinedWords: [(Double, String)]? = {
                    if last.words == nil && line.words == nil { return nil }
                    return (last.words ?? []) + (line.words ?? [])
                }()
                merged[merged.count - 1] = LyricLine(
                    t: last.t,
                    text: (last.text + " " + line.text)
                        .trimmingCharacters(in: .whitespaces),
                    words: joinedWords)
            } else {
                merged.append(line)
            }
        }
        return merged
    }

    // MARK: render

    /// The brat behaviour on the sleeve: the line starts huge and steps
    /// down in size as words arrive and crowd it, lowercase, left-anchored,
    /// lettered over the cover dimmed to a quarter with its own shadow.
    /// Layout is recomputed for exactly the words visible so far, which is
    /// what makes the shrink happen live instead of being pre-decided.
    static let ramp = 0.12
    static let dim: Float = 0.26

    private static var layoutCache: [String: ([String], Int)] = [:]
    /// The dimmed sleeve, computed once per track instead of 12k multiplies
    /// per tick.
    private static var dimCache: (key: String, px: [UInt8])? = nil
    /// Whole frames memoized on their visual state. Between word pops the
    /// picture does not change, and rendering an identical frame ten times a
    /// second was the lyric lag: every tick re-lettered, re-published, and
    /// dragged a full SwiftUI pass behind it. Now a tick that changes
    /// nothing costs a dictionary lookup, and equality upstream stops the
    /// republish entirely.
    private static var frameMemo: (key: String, px: [UInt8])? = nil

    static func render(sheet: [LyricLine], at t: Double,
                       adlibs: [(start: Double, end: Double, text: String)] = [],
                       over art: [UInt8], artKey: String = "",
                       ink: (UInt8, UInt8, UInt8)) -> [UInt8] {
        // what the foot sings right now
        let liveAdlib = adlibs.first { t >= $0.start && t < $0.end }?.text
        // the foot takes the rows the whole ad-lib needs (up to two), and
        // the first voice's region gives way above it: clipping the second
        // singer mid-yeah was worse than a smaller headline
        let footRows: [String] = liveAdlib.map { adlib in
            Array(PixelFont.wrap(letterableTokens(adlib).joined(separator: " "),
                                 maxWidth: 60, scale: 1).prefix(3))
        } ?? []
        let footTop = footRows.isEmpty ? 64 : 64 - (footRows.count * 9 - 2) - 1
        let regionH = footRows.isEmpty ? 60 : footTop - 3
        let base: [UInt8]
        if let c = dimCache, c.key == artKey, !artKey.isEmpty {
            base = c.px
        } else {
            base = art.map { UInt8(Float($0) * dim) }
            if !artKey.isEmpty { dimCache = (artKey, base) }
        }
        var px = base

        guard let idx = currentIndex(sheet, t) else {
            drawFoot(&px, footRows, at: footTop, ink: ink)
            return px
        }
        let line = sheet[idx]
        let (t0, text) = (line.t, line.text)
        let t1 = idx + 1 < sheet.count ? sheet[idx + 1].t : t0 + 6
        let tokens = letterableTokens(text)
        guard !tokens.isEmpty else {
            drawFoot(&px, footRows, at: footTop, ink: ink)
            return px
        }

        // timing runs on the FULL line so sizes never change the rhythm
        let n = tokens.count
        let borns = birthTimes(tokens: tokens, t0: t0, t1: t1, words: line.words)
        var visible = 0
        var newestK = 1.0
        for i in 0..<n where t >= borns[i] {
            visible = i + 1
            newestK = min(1.0, (t - borns[i]) / ramp)
        }
        guard visible > 0 else { return px }

        // the frame's whole visual identity: line, words shown, and the
        // newest word's ramp bucketed to its handful of distinct steps
        let kBucket = newestK >= 1 ? 9 : Int(newestK * 8)
        let frameKey = "\(artKey)|\(idx)|\(visible)|\(kBucket)|\((liveAdlib ?? "").hashValue)"
        if let memo = frameMemo, memo.key == frameKey { return memo.px }

        let key = "\(t0)|\(visible)|\(text.hashValue)|\(regionH)"
        let (rows, scale): ([String], Int)
        if let hit = layoutCache[key] {
            (rows, scale) = hit
        } else {
            let laid = layout(tokens: Array(tokens.prefix(visible)), regionH: regionH)
            if layoutCache.count > 256 { layoutCache.removeAll() }
            layoutCache[key] = laid
            (rows, scale) = laid
        }
        guard !rows.isEmpty else { return px }

        let lineH = 7 * scale + 2
        // centred in the region the foot has left standing, NOT the panel:
        // centring on 64 was exactly how the two voices ended up on top of
        // each other whenever the second one sang
        var y = max(1, (regionH - (rows.count * lineH - 2)) / 2)
        var drawn = 0
        for row in rows {
            let words = row.split(separator: " ").map(String.init)
            let x = 2
            // shadow: the row's own footprint, darkened
            var mask = [UInt8](repeating: 0, count: 64 * 64 * 3)
            PixelDraw.text(&mask, row, x: x + 1, y: y + 1,
                           rgb: (255, 255, 255), scale: scale)
            for i in stride(from: 0, to: mask.count, by: 3) where mask[i] > 0 {
                px[i] = UInt8(Float(px[i]) * 0.3)
                px[i + 1] = UInt8(Float(px[i + 1]) * 0.3)
                px[i + 2] = UInt8(Float(px[i + 2]) * 0.3)
            }
            drawn += words.count
            let isLastRow = drawn >= visible
            let settled = isLastRow && words.count > 0
                ? words.dropLast().joined(separator: " ") : row
            if !settled.isEmpty {
                PixelDraw.text(&px, settled, x: x, y: y, rgb: ink, scale: scale)
            }
            if isLastRow, let last = words.last {
                let lx = x + (settled.isEmpty ? 0
                    : PixelFont.textWidth(settled, scale: scale) + 6 * scale)
                let k = Float(newestK)
                let voiced = (UInt8(Float(ink.0) * max(0.3, k)),
                              UInt8(Float(ink.1) * max(0.3, k)),
                              UInt8(Float(ink.2) * max(0.3, k)))
                PixelDraw.text(&px, last, x: lx, y: y, rgb: voiced, scale: scale)
            }
            y += lineH
        }
        // the foot belongs to the second voice, and to no one else: when
        // nobody is answering, it stays dark
        drawFoot(&px, footRows, at: footTop, ink: ink)
        frameMemo = (frameKey, px)
        return px
    }

    /// The second singer's rows at the panel's foot: whole, never clipped
    /// mid-word, parentheses kept because that is how the second voice
    /// writes. Empty rows mean nobody is answering.
    private static func drawFoot(_ px: inout [UInt8], _ rows: [String],
                                 at top: Int, ink: (UInt8, UInt8, UInt8)) {
        guard !rows.isEmpty else { return }
        var y = top
        for row in rows {
            let x = (64 - min(60, PixelFont.textWidth(row, scale: 1))) / 2
            var mask = [UInt8](repeating: 0, count: 64 * 64 * 3)
            PixelDraw.text(&mask, row, x: x + 1, y: y + 1,
                           rgb: (255, 255, 255), scale: 1)
            for i in stride(from: 0, to: mask.count, by: 3) where mask[i] > 0 {
                px[i] = UInt8(Float(px[i]) * 0.35)
                px[i + 1] = UInt8(Float(px[i + 1]) * 0.35)
                px[i + 2] = UInt8(Float(px[i + 2]) * 0.35)
            }
            let k: Float = 0.66
            PixelDraw.text(&px, row, x: x, y: y,
                           rgb: (UInt8(Float(ink.0) * k),
                                 UInt8(Float(ink.1) * k),
                                 UInt8(Float(ink.2) * k)), scale: 1)
            y += 9
        }
    }

    /// Biggest type the visible words allow: one line if any size holds it,
    /// then whole-word wraps, then the small hard-wrapped truth.
    static func layout(tokens: [String], regionH: Int = 60) -> ([String], Int) {
        let joined = tokens.joined(separator: " ")
        for scale in [4, 3, 2, 1]
        where PixelFont.textWidth(joined, scale: scale) <= 60
            && 7 * scale <= regionH {
            return ([joined], scale)
        }
        for scale in [3, 2, 1] {
            let rows = PixelFont.wrap(joined, maxWidth: 60, scale: scale)
            if rows.count * (7 * scale + 2) - 2 <= regionH - 2,
               rows.joined(separator: " ") == joined {
                return (rows, scale)
            }
        }
        let maxRows = max(1, (regionH - 2) / 9)
        return (Array(PixelFont.wrap(joined, maxWidth: 60, scale: 1).suffix(maxRows)), 1)
    }

    /// When each token arrives: the sheet's own Enhanced-LRC stamps when it
    /// has them, else a spread weighted by word length, because long words
    /// take longer to sing and equal steps never felt like the song.
    static func birthTimes(tokens: [String], t0: Double, t1: Double,
                           words: [(Double, String)]?) -> [Double] {
        if let words {
            let stamped = words.compactMap { (wt, wtxt) -> Double? in
                let kept = PixelFont.normalize(wtxt.lowercased())
                    .filter { PixelFont.cell($0) != nil || $0 == " " }
                    .trimmingCharacters(in: .whitespaces)
                return kept.isEmpty ? nil : wt
            }
            if stamped.count == tokens.count { return stamped }
        }
        let span = min(2.4, max(0.6, (t1 - t0) * 0.55))
        let weights = tokens.map { Double(max(2, $0.count)) }
        let total = weights.reduce(0, +)
        var borns: [Double] = []
        var acc = 0.0
        for w in weights {
            borns.append(t0 + span * acc / total)
            acc += w
        }
        return borns
    }

    static func letterableTokens(_ text: String) -> [String] {
        PixelFont.normalize(text.lowercased())
            .filter { PixelFont.cell($0) != nil || $0 == " " }
            .split(separator: " ").map(String.init)
    }

    private static func currentIndex(_ sheet: [LyricLine], _ t: Double) -> Int? {
        guard let first = sheet.first, t >= first.t else { return nil }
        var lo = 0, hi = sheet.count - 1
        while lo < hi {
            let mid = (lo + hi + 1) / 2
            if sheet[mid].t <= t { lo = mid } else { hi = mid - 1 }
        }
        return lo
    }

    /// Wrapped rows of the letterable part of a line.
    private static func letterable(_ text: String) -> [String] {
        let kept = PixelFont.normalize(text)
            .filter { PixelFont.cell($0) != nil || $0 == " " }
            .split(separator: " ").joined(separator: " ")
        guard !kept.isEmpty else { return [] }
        // six rows is the panel's whole height at this leading; a line that
        // wants more gets its LAST six, because the end of a lyric line is
        // the part being sung when the cap matters
        return Array(PixelFont.wrap(kept, maxWidth: 60, scale: 1).suffix(6))
    }
}

File: brain/art/lyrics.py
"""Lyrics over the sleeve, in time with the song.

Source: LRCLIB (lrclib.net), the open synced-lyrics archive. One GET keyed
by track, artist, album and duration; the answer's syncedLyrics is LRC text,
"[mm:ss.xx] line" per line. No auth, no key, and misses are common enough
that having none is a first-class state here, never an error.

Display: the brat look. A flat field in the album's loudest voice, the
line in lowercase filling the panel from the left, black or white by
whatever the field needs, words popping in one at a time. Big where the
line is short, smaller only when it must be. Between lines the field just
holds, which is the confidence the whole aesthetic runs on. LRC gives line
times, not word times, so word timing is an even spread, which is a guess,
and an honest one: it is how people read.
"""
import colorsys
import re
import threading

import numpy as np
from PIL import Image

from .pixelfont import cell, draw_text, normalize, text_width
from .text_modes import wrap_text

_STAMP = re.compile(r"\[(\d+):(\d+(?:\.\d+)?)\]")


def parse_lrc(text: str) -> list[tuple[float, str]]:
    """[(seconds, line)], sorted. A line may carry several stamps."""
    out = []
    for raw in text.splitlines():
        stamps = _STAMP.findall(raw)
        if not stamps:
            continue
        words = _STAMP.sub("", raw).strip()
        for m, s in stamps:
            out.append((int(m) * 60 + float(s), words))
    out.sort(key=lambda p: p[0])
    # lines sharing a timestamp collapse into one, or the earlier of the
    # pair never shows at all
    merged: list[tuple[float, str]] = []
    for t, text in out:
        if merged and t - merged[-1][0] < 0.05:
            prev_t, prev = merged[-1]
            merged[-1] = (prev_t, (prev + " " + text).strip())
        else:
            merged.append((t, text))
    return merged


def split_voices(lines):
    """Two voices out of one sheet: parentheses are the second singer,
    wherever they sit in a line. Each ad-lib owns a window to the next
    event; what remains outside the parentheses is the first voice."""
    mains, raw = [], []
    for t, text in lines:
        text = text.strip()
        for g in re.findall(r"\([^()]*\)", text):
            if len(g) > 2:
                raw.append((t, g))
        main = " ".join(re.sub(r"\([^()]*\)", " ", text).split())
        if main:
            mains.append((t, main))
    starts = sorted(t for t, _ in lines)
    adlibs = []
    for t, text in raw:
        nxt = next((x for x in starts if x > t + 0.05), t + 4)
        adlibs.append((t, min(nxt, t + 6), text))
    return mains, adlibs


class LyricSheet:
    """One track's synced lines, and where we are in them."""

    def __init__(self, lines: list[tuple[float, str]]):
        self.lines, self.adlibs = split_voices(lines)

    def at(self, t: float):
        """(line_text, starts, ends) for time t; text may be "" between
        lines that LRC marks as instrumental gaps."""
        lines = self.lines
        if not lines or t < lines[0][0]:
            return "", 0.0, lines[0][0] if lines else 0.0
        lo, hi = 0, len(lines) - 1
        while lo < hi:
            mid = (lo + hi + 1) // 2
            if lines[mid][0] <= t:
                lo = mid
            else:
                hi = mid - 1
        t0 = lines[lo][0]
        t1 = lines[lo + 1][0] if lo + 1 < len(lines) else t0 + 6.0
        return lines[lo][1], t0, t1


def fetch_sheet(artist: str, title: str, album: str,
                duration_s: float | None) -> LyricSheet | None:
    """Exact lookup, distrusted when its edit's length disagrees with the
    song, then a duration-ranked search. Transport errors retry with a
    breath between attempts, so a hiccup is never remembered forever as
    'this song has no words'. None means none exist."""
    import time as _time

    import requests
    base = "https://lrclib.net/api"
    for attempt in range(3):
        if attempt:
            _time.sleep(2 * attempt)
        try:
            params = {"artist_name": artist, "track_name": title}
            if album:
                params["album_name"] = album
            if duration_s:
                params["duration"] = int(duration_s)
            resp = requests.get(base + "/get", params=params, timeout=10)
            data = resp.json() if resp.status_code == 200 else None
            if data and data.get("syncedLyrics") and duration_s \
                    and abs((data.get("duration") or duration_s)
                            - duration_s) > 4:
                data = None          # right song, wrong edit: keep looking
            if not (data and data.get("syncedLyrics")):
                resp = requests.get(base + "/search",
                                    params={"artist_name": artist,
                                            "track_name": title},
                                    timeout=10)
                hits = resp.json() if resp.status_code == 200 else []
                synced = [h for h in hits if h.get("syncedLyrics")]
                if duration_s and synced:
                    data = min(synced, key=lambda h: abs(
                        (h.get("duration") or 1e9) - duration_s))
                else:
                    data = synced[0] if synced else None
            if not (data and data.get("syncedLyrics")):
                return None
            lines = parse_lrc(data["syncedLyrics"])
            return LyricSheet(lines) if lines else None
        except Exception:
            continue
    return None


class LyricBook:
    """The fetch, made non-blocking: ask() starts a thread once per track and
    the render loop reads whatever state exists this frame."""

    def __init__(self):
        self.track = None
        self.sheet = None
        self.state = "idle"          # idle | loading | done | none

    def ask(self, track_id, artist, title, album, duration_s):
        if track_id == self.track:
            return
        self.track = track_id
        self.sheet = None
        self.state = "loading"

        def work():
            sheet = fetch_sheet(artist or "", title or "", album or "",
                                duration_s)
            # a slow answer for a track we already left is nobody's news
            if self.track == track_id:
                self.sheet = sheet
                self.state = "done" if sheet else "none"

        threading.Thread(target=work, daemon=True).start()


class LyricCanvas:
    """The brat behaviour on the sleeve: a line starts huge and steps down
    in size as its words arrive and crowd it, lowercase, left-anchored,
    lettered over the cover dimmed to a quarter, each row throwing its own
    shadow. Layout is recomputed for exactly the words visible so far."""

    DIM = 0.26
    RAMP = 0.12

    def __init__(self, size: int, art, sheet: LyricSheet,
                 color: str = "#f4f1ea"):
        self.size = size
        self.sheet = sheet
        if art is not None:
            self.base = np.asarray(
                art.convert("RGB").resize((size, size), Image.LANCZOS),
                dtype=np.float32) * self.DIM
        else:
            self.base = np.zeros((size, size, 3), dtype=np.float32)
        self.ink = (244, 241, 234)
        self._laid: dict[tuple, tuple] = {}

    def _tokens(self, text: str) -> list[str]:
        kept = "".join(
            ch for ch in normalize(text.lower())
            if cell(ch) is not None or ch == " ")
        return kept.split()

    def _layout(self, key, tokens: list[str], region_h: int = 60):
        hit = self._laid.get(key)
        if hit is not None:
            return hit
        joined = " ".join(tokens)
        rows, scale = None, 1
        for s_ in (4, 3, 2, 1):
            if text_width(joined, s_) <= self.size - 4 and 7 * s_ <= region_h:
                rows, scale = [joined], s_
                break
        if rows is None:
            for s_ in (3, 2, 1):
                cand = wrap_text(joined, self.size - 4, s_)
                if (len(cand) * (7 * s_ + 2) - 2 <= region_h - 2
                        and " ".join(cand) == joined):
                    rows, scale = cand, s_
                    break
        if rows is None:
            max_rows = max(1, (region_h - 2) // 9)
            rows, scale = wrap_text(joined, self.size - 4, 1)[-max_rows:], 1
        if len(self._laid) > 256:
            self._laid.clear()
        self._laid[key] = (rows, scale)
        return rows, scale

    def _foot_rows(self, text: str) -> list[str]:
        tokens = self._tokens(text)
        if not tokens:
            return []
        return wrap_text(" ".join(tokens), self.size - 4, 1)[:3]

    def _foot(self, canvas, rows: list[str], top: int):
        """The second singer's rows at the panel's foot: whole, never
        clipped mid-word, parentheses kept because that is how the second
        voice writes."""
        y = top
        for row in rows:
            x = (self.size - min(self.size - 4, text_width(row, 1))) // 2
            u8 = np.zeros((self.size, self.size, 3), dtype=np.uint8)
            draw_text(u8, row, x + 1, y + 1, (255, 255, 255), 1)
            mask = u8.sum(axis=2) > 0
            canvas[mask] *= 0.35
            foot_ink = tuple(int(c * 0.66) for c in self.ink)
            u8[:] = 0
            draw_text(u8, row, x, y, foot_ink, 1)
            mask = u8.sum(axis=2) > 0
            canvas[mask] = u8[mask]
            y += 9

    def frame_at(self, t: float) -> Image.Image:
        canvas = self.base.copy()
        text, t0, t1 = self.sheet.at(t)
        # the foot row's claimants: a live ad-lib outranks the preview
        adlib = next((a[2] for a in self.sheet.adlibs
                      if a[0] <= t < a[1]), None)
        foot_rows = self._foot_rows(adlib) if adlib else []
        foot_top = self.size if not foot_rows \
            else self.size - (len(foot_rows) * 9 - 2) - 1
        region_h = 60 if not foot_rows else foot_top - 3
        tokens = self._tokens(text) if text else []
        if tokens:
            n = len(tokens)
            span = min(2.4, max(0.6, (t1 - t0) * 0.55))
            step = span / n
            visible, newest_k = 0, 1.0
            for i in range(n):
                born = t0 + i * step
                if t >= born:
                    visible = i + 1
                    newest_k = min(1.0, (t - born) / self.RAMP)
            if visible:
                rows, scale = self._layout((t0, visible, region_h),
                                           tokens[:visible], region_h)
                line_h = 7 * scale + 2
                y = max(1, (region_h - (len(rows) * line_h - 2)) // 2)
                drawn = 0
                for row in rows:
                    words = row.split()
                    # the row's shadow, then its ink
                    u8 = np.zeros((self.size, self.size, 3), dtype=np.uint8)
                    draw_text(u8, row, 3, y + 1, (255, 255, 255), scale)
                    mask = u8.sum(axis=2) > 0
                    canvas[mask] *= 0.3
                    drawn += len(words)
                    last_row = drawn >= visible
                    settled = " ".join(words[:-1]) if last_row else row
                    if settled:
                        u8[:] = 0
                        draw_text(u8, settled, 2, y, self.ink, scale)
                        mask = u8.sum(axis=2) > 0
                        canvas[mask] = u8[mask]
                    if last_row and words:
                        lx = 2 + (text_width(settled, scale) + 6 * scale
                                  if settled else 0)
                        k = max(0.3, newest_k)
                        ink = tuple(int(c * k) for c in self.ink)
                        u8[:] = 0
                        draw_text(u8, words[-1], lx, y, ink, scale)
                        mask = u8.sum(axis=2) > 0
                        canvas[mask] = u8[mask]
                    y += line_h
        if foot_rows:
            self._foot(canvas, foot_rows, foot_top)
        return Image.fromarray(np.clip(canvas, 0, 255).astype(np.uint8), "RGB")
